pkg/streamingpromql/optimize/plan/commonsubexpressionelimination/series_data_ring_buffer.go

```diff
@@ -0,0 +1,90 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package commonsubexpressionelimination
+
+import (
+	"fmt"
+
+	"github.com/grafana/mimir/pkg/streamingpromql/types"
+)
+
+type SeriesDataRingBuffer struct {
+	data []types.InstantVectorSeriesData
+
+	startIndex       int // Position in data of the first series.
+	firstSeriesIndex int // Series index of the first series in the buffer.
+	seriesCount      int // Number of series in the buffer.
+}
+
+func (b *SeriesDataRingBuffer) Append(d types.InstantVectorSeriesData, seriesIndex int) {
+	if b.seriesCount > 0 && seriesIndex != b.firstSeriesIndex+b.seriesCount {
+		panic(fmt.Sprintf("attempted to append series with index %v, but first series index in buffer is %v and have %v series", seriesIndex, b.firstSeriesIndex, b.seriesCount))
+	}
+
+	if len(b.data) == b.seriesCount {
+		// Buffer is full, need to grow it.
+		newSize := max(len(b.data)*2, 2) // Grow by powers of two, and ensure we have at least 2 slots.
+		newData := make([]types.InstantVectorSeriesData, newSize)
+		copy(newData, b.data[b.startIndex:])
+		copy(newData[len(b.data)-b.startIndex:], b.data[:b.startIndex])
+		b.startIndex = 0
+		b.data = newData
+	}
+
+	b.data[(b.startIndex+b.seriesCount)%len(b.data)] = d
+	b.seriesCount++
+
+	if b.seriesCount == 1 {
+		// Just added the first series.
```

Comment:
(nit) it feels more logical to me to have this at the start of the Append (and checking if it is 0). The first if block could then be an `else if` saving an extra check.

Reply:
Good point, changed in db6b4de